subprojects/core-api/src/main/java/org/gradle/api/tasks/PropertyNormalizer.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.tasks;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A normalizer used to remove unwanted noise when considering file inputs.
+ *
+ * @since 4.3
+ */
+@Incubating
+public interface PropertyNormalizer {
```

Comment:
I don't quite like the name `PropertyNormalizer`. I think it should say that it has to do something with files. How about `FileCollectionNormalizer`? I then would just drop the `Property` from the other normalizers: `ClasspathNormalizer`, `CompileClasspathNormalizer` and `GenericFileCollectionNormalizer`. We could also keep the `Property`, so e.g. `FilePropertyNormalizer`, `ClasspathPropertyNormalizer` and `GenericFilePropertyNormalizer`. WDYT?

Reply:
I decided against `FileCollectionNormalizer` because it doesn't have to be used on a `FileCollection`: you could apply it to a `File` (even if it's not pointing to a directory), of even a string.

I'm okay with `PropertyFileNormalizer`.

Reply:
Hm... Considering that these normalizers might be used in worker API caching, too, we should probably not use the name `Property` either. What about `FileNormalizer`?

Reply:
`FileNormalizer` sounds good to me! Do we want a better name for `GenericFileNormalizer`? I don't even know if we should expose the type on the public API given that you would mostly use `withPathSensitivity` instead of using `GenericFileNormalizer`.

Reply:
Okay, let's go with `FileNormalizer`.

I'm also on the fence about `GenericFileNormalizer`. I guess we should keep it internal for now, and expose it if there is an expressed need.

Reply:
Done. BTW, we haven't really future-proofed anything by dropping `PropertyNormalizer`, as `FileNormalizer` is still in the `org.gradle.api.tasks` package, but I guess that's fine for now.